aten/src/ATen/native/cuda/RangeFactories.cu

```diff
@@ -31,18 +31,29 @@ Tensor& linspace_cuda_out(Tensor& result, Scalar start, Scalar end, int64_t step
     // skip
   } else if (steps == 1) {
     r.fill_(start);
+  } else if (isComplexType(r.scalar_type())) {
+    AT_DISPATCH_COMPLEX_TYPES(at::ScalarType::Half, r.scalar_type(), "linspace_cuda", [&]() {
+      scalar_t scalar_start = start.to<scalar_t>();
+      scalar_t scalar_end = end.to<scalar_t>();
+      scalar_t step = (scalar_end - scalar_start) / static_cast<scalar_t>(steps - 1);
+
+      auto iter = TensorIterator::nullary_op(r);
+      gpu_kernel_with_index(iter, [scalar_start, step]GPU_LAMBDA(int ind) -> scalar_t {
+        scalar_t val = scalar_start + step * static_cast<scalar_t>(ind);
+        return val;
+      });
+    });
   } else {
     AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Half, r.scalar_type(), "linspace_cuda", [&]() {
       scalar_t scalar_start = start.to<scalar_t>();
       scalar_t scalar_end = end.to<scalar_t>();
-      scalar_t step = (scalar_end - scalar_start) / static_cast<scalar_t>(steps - 1);
+      double step = static_cast<double>(scalar_end - scalar_start) / (steps - 1);
```

Comment:
@enearlor can you please leave a comment here about why double is needed (is it?). Operations with doubles are typically much slower on cuda, so unless absolutely necessary we try not to use them, and it seems like in most cases linspace should be ok with a float step. I know you've ran many experiments, so I'm curious to hear which cases fail with floats.

Reply:
I chose to use `double` after finding that I was having issues with the numerical accuracy of `step` when `linspace` is dispatched to `float`. The inaccuracy is small, and is typically around the type's `eps`. I reasoned that the discrepancy is due to Python doing everything in higher precision by default, and then rounding at the end. Replicating this in `linspace` fixed the discrepancy.

I considered the performance hit this might cause, but I have not been able to investigate it as I can't directly test the CUDA build. I reasoned that the change in overall clock time would be minor as `linspace` is not frequently recalled. If this is a concern, then I can change it back. I list my suggestions below.

At the moment, these are the solutions I'm considering.
* Leave the code as-is, other then the corrections for test_linspace and the XLA build. This ideally results in the CUDA and CPU versions giving the same results.
* Change the type back to `scalar_t`. This would leave minor numerical discrepancies between the output of the CUDA and CPU version, but it would keep the CUDA version performing at the speed it is currently. This would require me to separate out the Integral form though, as an integral `step` can be truncated to 0. This would also mean that I'd have to allow for error between the expected result and actual result in the tests, but I believe that is the current state of affairs as well.

Reply:
I'd prefer changing the type back to `scalar_t` and increasing test tolerance if needed. Separating integral dispatch is fine. Thank you!

Reply:
Okay. I added a commit to address this. The integral dispatch still uses `float` for the `step`, but otherwise uses `scalar_t`.

To verify, is the CPU version as is? It relies on declaring the step to be `double` type, as this did prior to the commit.

Reply:
Yes, for CPU doubles don't carry such a performance penalty. XLA failures are probably unrelated, can you try rebasing and see it they are fixed?

Reply:
The XLA failures appear to be fixed after rebasing. The onnx build is failing for potentially similar reasons.